cpp/src/arrow/compute/exec/exec_plan.h

```diff
@@ -144,10 +146,24 @@ class ARROW_EXPORT ExecPlan : public std::enable_shared_from_this<ExecPlan> {
   /// \brief Return the plan's attached metadata
   std::shared_ptr<const KeyValueMetadata> metadata() const;
 
+  /// \brief Should the plan use a legacy batching strategy
+  ///
+  /// This is currently in place only to support the Scanner::ToTable
+  /// method.  This method relies on batch indices from the scanner
+  /// remaining consistent.  This is impractical in the ExecPlan which
+  /// might slice batches as needed (e.g. for a join)
+  ///
+  /// However, it still works for simple plans and this is the only way
+  /// we have at the moment for maintaining implicit order.
+  bool UseLegacyBatching() const { return use_legacy_batching_; }
```

Comment:
I'm not super enthused about this flag, but it seems it'll be hard to get around it. Is there a plan for being able to get rid of it? 

Also, ScanBatches is the underlying method that requires sequencing right?

Reply:
In principle is it not possible to renumber the batches?

Reply:
I would very much like to properly address ordering soon (and I think @icexelloss needs it as well).  Once we do that then this flag can go away.  The scanner will assign the initial "sort index" to batches and the sink will just resequence based on this at the end.  The "sort index" will be a property of the batch itself (and there will not be one index series per fragment) and so it will be very easy for the source node to renumber things as it slices batches.

Renumbering today is possible, but not very straightforward.  The scanner adds a scalar column to the batch with the batch index.  So to renumber the batches I'd have to pull out the scalar column (which might not be present if the source is not a scanner), then reattach a scalar column to each sliced batch.  Another complication is that I would have to keep track of the current batch counter for each fragment.  All of this is possible, but enough of a change, that I'd rather not make it and instead focus on the better long term fix.

Reply:
> Also, ScanBatches is the underlying method that requires sequencing right?

ToTable uses `ScanBatchesUnorderedAsync` today and then inserts the items into a priority queue when they come out of Acero.  `ScanBatches` is nearly unused but I think there is still one python method that binds to it.

Reply:
> Once we do that then this flag can go away. The scanner will assign the initial "sort index" to batches and the sink will just resequence based on this at the end.

How will this interact with group by/union/join?